### DISCUSSION TITLE
Gotchas and pain points of react-select

### DISCUSSION THREAD

**Discussion by ebonow** [JedWatson/react-select#4312] [Q&A]
I'm hoping to compile an FAQ, so I wanted to ask the community what things regarding react-select were/are frustrating from an implementation or understanding standpoint?

I'll start with some of the things I struggled/struggle with in no specific order:

- **Upgrade guide**
  - Documentation provides upgrade path from V1 to V2 but does not mention V3 which seems confusing. 
  - Consolidating breaking changes from major version updates would make things more clear and help reduce frustration of missing features

- **Prop documentation/DOM hierarchy**
  - Common props for each custom component drown out the unique/noteworthy props
  - No clarification on which components require an innerRef if replaced
  - While very well articulated in the components section, a visual representation/diagram may help better explain the hierarchy for those understanding how to create their own custom components
  - No styling or classNamePrefix applied to dummyInput
  - No warnings or explanations for impact of accidentally re-rendering new ValueContainer or Input (lose focus while typing)

- **Using the styles api**
  - Requires background with styled components especially regarding nesting or pseudo selectors
  - Confusing when trying to style the input component
  - Input styles api does not return same kind of common state/props as other components
  - Learning Emotion to better understand the css prop passed to each component

- **Hiding input and clearing inputValue onChange/onBlur**
  - Confusing behavior for new developers and end-users with little to no elaboration on what/why is happening regarding Placeholder, Value, and Input positioning and visibility.
  - Created discussion [discussing this more in depth](https://github.com/JedWatson/react-select/discussions/4302) hoping to get more feedback and thoughts from others.

- **value = Object**
  - I understand why it is, but in practice, it is somewhat confusing that I am ultimately interested in not the `value` but the `value.value`

- **Accessing select ref from Async, Createable, and AsyncCreateable**
  - for example, say you wanted to get the inputRef of the Select
``` JS 
const inputRef = selectRef.current.select.inputRef;
const asyncInputRef = asynctRef.current.select.select.inputRef;
const createableInputRef = createableRef.current.select.select.inputRef;
const asyncCreatableInputRef = asyncCreatable.current.select.select.select.inputRef;
```

- **Hiding and Showing the menu**
  - Automated integration testing cannot easily access the menu without certain props being set (perhaps this and animation are a case for hiding the menu instead of render/not render) 

  - Without menu open/close methods, the menuIsOpen prop is a "great power/great responsibility" situation where you're allotted both more control and more complexity as it's not entirely clear in documentation how to handle this especially since many simply want to trigger without being responsible for understanding all of the close-menu conditions (onBlur, onChange, onClear, escape key, etc...). It's also not clear to many that closing it should be done by setting it to undefined and not false.

``` JS
const [ menuIsOpen, setMenuIsOpen ] = useState();
const openMenu = setMenuIsOpen(true);
const closeMenu = setMenuIsOpen(); // setting to false would prevent window from opening on focus or click

return <Select menuIsOpen={menuIsOpen} />
```

- **Input Props**: One of the more requested features/issues are for certain input props
  - maxLength, required, autoComplete, etc... can be accomplished by creating a custom Input, but is not well explained/documented

``` JS 
const Input = (props) => {
  const { maxLength, required } = props.selectProps;
  return <components.Input {...props} maxLength={maxLength} required={required} />
}

const styles = {
  input: css => ({
    'input:invalid': {
      opacity: '1 !important'
    }
  })
};

const MySelect =({ maxLength: 5, required: true, ...props }) => 
  <Select {...props} styles={styles} components={{Input}} maxLength={maxLength} required={required} />
```

I'm sure there's more that I am forgetting, but curious what other frustrations others have run into in the interest of lowering the learning curve for everyone.

**Comment by dolie** [comment]
missing required attribute is such a pain...

**Comment by joeyfigaro** [comment]
Better/complete documentation on an action's `meta` would be great. Looks like the [current docs only touch on action types and not the associated payloads](https://www.google.com/search?q=site%3Areact-select.com+meta&ei=P-VzY5rNNoWr5NoPo9qUuAE&ved=0ahUKEwja6Mjm8rD7AhWFFVkFHSMtBRcQ4dUDCBA&uact=5&oq=site%3Areact-select.com+meta&gs_lcp=Cgxnd3Mtd2l6LXNlcnAQA0oECEEYAUoECEYYAFDSAli2BGC8BmgBcAB4AIABN4gBZ5IBATKYAQCgAQHAAQE&sclient=gws-wiz-serp)?
